### PR TITLE
Convert SearchResult, InteractiveAuth, PushProcessor and Scheduler to TS

### PR DESCRIPTION
**Pull request by t3chguy**

### CODE REVIEW

**Comment by t3chguy** [comment]
Blocked until after RC

File: src/@types/PushRules.ts [hunk 1]
```diff
@@ -90,8 +90,7 @@ export interface ISenderNotificationPermissionCondition
     key: string;
 }
 
-export type PushRuleCondition = IPushRuleCondition<string>
-    | IEventMatchCondition
+export type PushRuleCondition = IEventMatchCondition
```
Comment (turt2live) [inline src/@types/PushRules.ts:93]: custom conditions are still possible though?

File: src/interactive-auth.ts [hunk 2]
```diff
@@ -510,12 +578,13 @@ InteractiveAuth.prototype = {
...
         err.name = 'NoAuthFlowFoundError';
-        err.required_stages = [];
-        if (haveEmail) err.required_stages.push(EMAIL_STAGE_TYPE);
-        if (haveMsisdn) err.required_stages.push(MSISDN_STAGE_TYPE);
-        err.available_flows = flows;
+        const requiredStages: string[] = [];
+        if (haveEmail) requiredStages.push(EMAIL_STAGE_TYPE);
+        if (haveMsisdn) requiredStages.push(MSISDN_STAGE_TYPE);
+        (err as any).required_stages = requiredStages;
+        (err as any).available_flows = flows;
```
Comment (turt2live) [inline src/interactive-auth.ts:585]: we should type the error, or at least leave comments about the cast. `// TODO: Types` is fine imo

File: src/pushprocessor.ts [hunk 3]
```diff
@@ -17,12 +16,46 @@ limitations under the License.
...
-const RULEKINDS_IN_ORDER = ['override', 'content', 'room', 'sender', 'underride'];
+const RULEKINDS_IN_ORDER = [
+    PushRuleKind.Override,
+    PushRuleKind.ContentSpecific,
+    PushRuleKind.RoomSpecific,
+    PushRuleKind.SenderSpecific,
+    PushRuleKind.Underride,
+];
+
+interface IRuleset {
```
Comment (turt2live) [inline src/pushprocessor.ts:48]: type already exists as `PushRuleSet`.

File: src/pushprocessor.ts [hunk 4]
```diff
@@ -17,12 +16,46 @@ limitations under the License.
...
+    [PushRuleKind.ContentSpecific]: IPushRule[];
+    [PushRuleKind.Override]: IPushRule[];
+    [PushRuleKind.RoomSpecific]: IPushRule[];
+    [PushRuleKind.SenderSpecific]: IPushRule[];
+    [PushRuleKind.Underride]: IPushRule[];
+}
+
+export interface IRulesets {
```
Comment (turt2live) [inline src/pushprocessor.ts:56]: type already exists

File: src/pushprocessor.ts [hunk 5]
```diff
@@ -64,31 +97,97 @@ const DEFAULT_OVERRIDE_RULES = [
...
+export class PushProcessor {
+    /**
+     * Convert a list of actions into a object with the actions as keys and their values
+     * eg. [ 'notify', { set_tweak: 'sound', value: 'default' } ]
+     *     becomes { notify: true, tweaks: { sound: 'default' } }
+     * @param {array} actionList The actions list
+     *
+     * @return {object} A object with key 'notify' (true or false) and an object of actions
+     */
+    public static actionListToActionsObject(actionList: PushRuleAction[]): IActionsObject {
```
Comment (turt2live) [inline src/pushprocessor.ts:125]: please can we put static functions below the constructor :(

File: src/scheduler.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        }
+        return queue.shift();
+    }
+}
+
+function debuglog(...args) {
+    if (DEBUG) {
+        logger.log(...args);
+    }
+}
```
Comment (turt2live) [inline src/scheduler.ts:291]: one day we should just use the logging library directly.

**Review by turt2live — CHANGES_REQUESTED** [review]

**Comment by turt2live** [comment]
gah, pressed buttons too quickly: overall the first readthrough of the diff looks fine. It's large enough where there's probably going to be a second read on re-review and a small chance of a third.

File: src/@types/PushRules.ts (hunk 1 above)
Comment (turt2live) [reply src/@types/PushRules.ts:93]: discussed oob: removing this type makes life easier down the road, and realistically we don't care about custom conditions at this layer. Would be fine with a comment acknowledging this.

**Comment by t3chguy** [comment]
@turt2live please merge after approval along with https://github.com/matrix-org/matrix-react-sdk/pull/6323 - I will be on holiday

**Review by turt2live — APPROVED** [review]

**Pull request merged**